optics-core/src/Optics/Internal/Indexed/Classes.hs

```diff
@@ -0,0 +1,514 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_HADDOCK not-home #-}
+
+-- | Internal implementation details of indexed optics.
+--
+-- This module is intended for internal use only, and may change without warning
+-- in subsequent releases.
+module Optics.Internal.Indexed.Classes where
+
+import Control.Applicative
+import Control.Applicative.Backwards
+import Control.Monad.Trans.Identity
+import Control.Monad.Trans.Reader
+import Data.Functor.Constant
+import Data.Functor.Compose
+import Data.Functor.Identity
+import Data.Functor.Product
+import Data.Functor.Reverse
+import Data.Functor.Sum
+import Data.Ix
+import Data.List.NonEmpty
+import Data.Monoid hiding (Product, Sum)
+import Data.Proxy
+import Data.Tree
+import Data.Void
+import GHC.Generics
+import qualified Data.Array as Array
+import qualified Data.IntMap as IntMap
+import qualified Data.Map as Map
+import qualified Data.Sequence as Seq
+
+import Optics.Internal.Utils
+
+-- | Class for 'Functor's that have an additional read-only index available.
+class Functor f => FunctorWithIndex i f | f -> i where
+  imap :: (i -> a -> b) -> f a -> f b
+  default imap
+    :: TraversableWithIndex i f => (i -> a -> b) -> f a -> f b
+  imap f = runIdentity #. itraverse (\i a -> Identity (f i a))
+  {-# INLINE imap #-}
+
+-- | Class for 'Foldable's that have an additional read-only index available.
+class (FunctorWithIndex i f, Foldable f
+      ) => FoldableWithIndex i f | f -> i where
+  ifoldMap :: Monoid m => (i -> a -> m) -> f a -> m
+  default ifoldMap
+    :: (TraversableWithIndex i f, Monoid m) => (i -> a -> m) -> f a -> m
+  ifoldMap f = getConst #. itraverse (\i a -> Const (f i a))
+  {-# INLINE ifoldMap #-}
+
+  ifoldr :: (i -> a -> b -> b) -> b -> f a -> b
+  ifoldr iabb b0 = (\e -> appEndo e b0) . ifoldMap (\i -> Endo #. iabb i)
+  {-# INLINE ifoldr #-}
+
+  ifoldl' :: (i -> b -> a -> b) -> b -> f a -> b
+  ifoldl' ibab b0 s = ifoldr (\i a bb b -> bb $! ibab i b a) id s b0
+  {-# INLINE ifoldl' #-}
```

Comment:
I wonder if `itoList` should be the member class.

```haskell
instance FoldableWithIndex k (Map.Map k v) where
    itoList = Map.toList
```

Not an issue for this PR, more of a note for #309

Reply:
Makes sense, I presume we'd want to mimic `Foldable` with the amount of methods if it's to be split.

Reply:
Yes, it might make sense to add all four of `ifoldr`, `ifoldr'`, `ifoldl'`, and `ifoldl`. And then have `ifoldMap` and `itoList`, so six in total.